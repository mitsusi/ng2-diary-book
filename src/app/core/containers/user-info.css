.image > img {
  width: 32px;
  height: 32px;
  border-radius: 50%;
}

.user-info {
  display: flex;
  flex-direction: row;
  align-self: center;
  align-items: center;
}

.name {
  font-size: small;
}
