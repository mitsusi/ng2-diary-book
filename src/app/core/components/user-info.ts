import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'app-user-info-component',
  template: `
    <div class="user-info">
      <div class="image" *ngIf="hasPhotoURL()">
        <img src="{{photoURL}}" />
      </div>
      <div class="name">
        <span>{{displayName}}</span>
      </div>
      <span class="exit-button">
        <md-icon (click)="onExit.emit();">exit_to_app</md-icon>
      </span>
    </div>
  `,
  styleUrls: ['./user-info.css'],
})
export class UserInfoComponent {
  @Input() displayName: string;
  @Input() photoURL: string;
  @Output() onExit = new EventEmitter();

  hasPhotoURL() {
    return this.photoURL !== '';
  }
}
