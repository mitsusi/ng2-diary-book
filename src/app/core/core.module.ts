import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';

import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { AuthInterceptor } from './modules/http-interceptors/auth-append-token';

import {
  MatInputModule,
  MatButtonModule,
  MatIconModule,
  MatDatepickerModule,
  MatNativeDateModule,
  MatCardModule,
  MatSidenavModule,
  MatListModule,
  MatToolbarModule,
  MatProgressSpinnerModule,
} from '@angular/material';

import { AppComponent } from './containers/app';
import { NotFoundPageComponent } from './containers/not-found-page';
import { LayoutComponent } from './components/layout';
import { NavItemComponent } from './components/nav-item';
import { SidenavComponent } from './components/sidenav';
import { ToolbarComponent } from './components/toolbar';
import { NotLoggedInComponent } from './components/not-logged-in';
import { CommonShowIfLoggedInComponent } from './containers/common-show-if-logged-in';
import { LoaderComponent } from './containers/loader';
import { BreadcrumbsComponent } from './components/breadcrumbs';
import { BreadcrumbsContainer } from './containers/breadcrumbs';
import { UserInfoContainer } from './containers/user-info';

import { BreadcrumbsService } from './services/breadcrumbs';

import { PipesModule } from '../shared/pipes';
import { EffectsModule } from '@ngrx/effects';
import { RouterEffects } from './effects/router';
import { FormsModule } from '@angular/forms';
import { TagsAutoCompleteModule } from './features/components/tags-autocomplete/tags-autocomplete.module';
import { TagsViewerModule } from './features/components/tags-viewer/tags-viewer.module';
import { RouterHelperService } from './services/router';
import { UserInfoComponent } from './components/user-info';
import { AuthService } from './services/auth.service';

export const COMPONENTS = [
  AppComponent,
  NotFoundPageComponent,
  LayoutComponent,
  NavItemComponent,
  SidenavComponent,
  ToolbarComponent,
  NotLoggedInComponent,
  CommonShowIfLoggedInComponent,
  LoaderComponent,
  BreadcrumbsComponent,
  BreadcrumbsContainer,
  UserInfoContainer,
  UserInfoComponent,
];

const MAT_COMPONENTS = [
  MatInputModule,
  MatButtonModule,
  MatIconModule,
  MatDatepickerModule,
  MatNativeDateModule,
  MatCardModule,
  MatSidenavModule,
  MatListModule,
  MatToolbarModule,
  MatProgressSpinnerModule,
];

@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    RouterModule,
    PipesModule,
    HttpClientModule,
    EffectsModule.forRoot([RouterEffects]),
    TagsAutoCompleteModule,
    TagsViewerModule,
    [...MAT_COMPONENTS],
  ],
  declarations: COMPONENTS,
  exports: [
    ...COMPONENTS,
    CommonModule,
    FormsModule,
    HttpClientModule,
    TagsAutoCompleteModule,
    TagsViewerModule,
    [...MAT_COMPONENTS],
  ],
  providers: [
    RouterHelperService,
    BreadcrumbsService,
    {
      provide: HTTP_INTERCEPTORS,
      useClass: AuthInterceptor,
      multi: true,
    },
    AuthService,
  ],
})
export class CoreModule {
  static forRoot() {
    return {
      ngModule: CoreModule,
    };
  }
}
