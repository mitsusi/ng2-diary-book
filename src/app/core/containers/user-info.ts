import { ChangeDetectorRef, Component } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Store } from '@ngrx/store';
import { AsyncPipe } from '@angular/common';

import * as fromRoot from '../../reducers';

@Component({
  selector: 'app-user-info',
  template: `
    <app-user-info-component
      [photoURL]="photoURL$ | async"
      [displayName]="displayName$ | async"
      (onExit)="onExitClick();"
    ></app-user-info-component>
  `,
})
export class UserInfoContainer {
  displayName$: Observable<string>;
  photoURL$: Observable<string>;

  constructor(private store: Store<fromRoot.State>) {
    store
      .select(fromRoot.getUserIsLoggedIn)
      .filter(isUserLoggedIn => isUserLoggedIn)
      .subscribe(() => {
        this.displayName$ = store.select(fromRoot.getDisplayName);
        this.photoURL$ = store.select(fromRoot.getPhotoURL);
      });
  }

  onExitClick(): void {
    console.log('exit');
  }
}
