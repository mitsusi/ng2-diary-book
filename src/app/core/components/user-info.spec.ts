/* tslint:disable:no-unused-variable */
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { UserInfoComponent } from './user-info';
import { MdIconModule } from '@angular/material';
import { By } from '@angular/platform-browser';
import { DebugElement } from '@angular/core';

describe('UserInfoComponent', () => {
  let component: UserInfoComponent;
  let fixture: ComponentFixture<UserInfoComponent>;

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        imports: [MdIconModule],
        declarations: [UserInfoComponent],
      }).compileComponents();
    }),
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(UserInfoComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('should handle photoURL correctly', () => {
    describe('photoURL is passed', () => {
      beforeEach(() => {
        component.photoURL = 'somePhoto';
        fixture.detectChanges();
      });
      it('hasPhotoURL should be truthy', () =>
        expect(component.hasPhotoURL()).toBeTruthy());
      it('should match snapshot', () =>
        expect(fixture.nativeElement.innerHTML).toMatchSnapshot());
    });
    describe('photoURL is not passed', () => {
      beforeEach(() => {
        component.photoURL = '';
        fixture.detectChanges();
      });
      it('hasPhotoURL should be falsy', () =>
        expect(component.hasPhotoURL()).toBeFalsy());
      it('should match snapshot', () =>
        expect(fixture.nativeElement.innerHTML).toMatchSnapshot());
    });
  });

  describe('should handle displayName correctly', () => {
    it('should show display name if it is passed', () => {
      component.displayName = 'vova chuma';
      fixture.detectChanges();

      expect(fixture.nativeElement.innerHTML).toMatchSnapshot();
    });
  });

  it('should handle click on exit button', () => {
    const spy = jest.fn();
    component.onExit.subscribe(spy);

    const de: DebugElement = fixture.debugElement.query(
      By.css('span > md-icon'),
    );

    de.triggerEventHandler('click', null);

    expect(spy).toHaveBeenCalled();
  });
});
