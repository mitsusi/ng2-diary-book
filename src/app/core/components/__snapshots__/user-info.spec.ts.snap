// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfoComponent should handle displayName correctly should show display name if it is passed 1`] = `
"
    <div class=\\"user-info\\">
      <!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><div class=\\"image\\">
        <img src=\\"\\">
      </div>
      <div class=\\"name\\">
        <span>vova chuma</span>
      </div>
      <span class=\\"exit-button\\">
        <md-icon class=\\"mat-icon material-icons\\" role=\\"img\\" aria-hidden=\\"true\\">exit_to_app</md-icon>
      </span>
    </div>
  "
`;

exports[`UserInfoComponent should handle photoURL correctly photoURL is not passed should match snapshot 1`] = `
"
    <div class=\\"user-info\\">
      <!--bindings={
  \\"ng-reflect-ng-if\\": \\"false\\"
}-->
      <div class=\\"name\\">
        <span></span>
      </div>
      <span class=\\"exit-button\\">
        <md-icon class=\\"mat-icon material-icons\\" role=\\"img\\" aria-hidden=\\"true\\">exit_to_app</md-icon>
      </span>
    </div>
  "
`;

exports[`UserInfoComponent should handle photoURL correctly photoURL is passed should match snapshot 1`] = `
"
    <div class=\\"user-info\\">
      <!--bindings={
  \\"ng-reflect-ng-if\\": \\"true\\"
}--><div class=\\"image\\">
        <img src=\\"somePhoto\\">
      </div>
      <div class=\\"name\\">
        <span></span>
      </div>
      <span class=\\"exit-button\\">
        <md-icon class=\\"mat-icon material-icons\\" role=\\"img\\" aria-hidden=\\"true\\">exit_to_app</md-icon>
      </span>
    </div>
  "
`;
